controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/ResourceMeterMaintainer.java

```diff
@@ -150,14 +150,21 @@ private void reportResourceSnapshots(Collection<ResourceSnapshot> resourceSnapsh
                 .filter(this::unlessNodeOwnerIsSystemApplication)
                 .filter(this::isNodeStateMeterable)
                 .filter(this::isClusterTypeMeterable)
+                // Grouping by ApplicationId -> Architecture -> ResourceSnapshot
                 .collect(Collectors.groupingBy(node ->
                                 node.owner().get(),
-                                Collectors.collectingAndThen(Collectors.toList(),
-                                        nodeList -> ResourceSnapshot.from(
-                                                nodeList,
-                                                clock.instant(),
-                                                zoneId))
-                                )).values();
+                                Collectors.collectingAndThen(Collectors.groupingBy(node -> node.resources().architecture()),
+                                        nodeList -> nodeList.entrySet()
+                                                .stream()
+                                                .collect(Collectors.toMap(
+                                                        Map.Entry::getKey,
+                                                        entry -> ResourceSnapshot.from(entry.getValue(), clock.instant(), zoneId))
+                                                ))
+                                ))
```

Comment:
When combining collectors in this way it can help readability to assign them to a variable when you create them and use that later. It makes the code less noisy and gives the collector a semantic name.